File: loopme-sdk/src/test/java/com/loopme/request/RequestValidatorTest.java
package com.loopme.request;

import static org.junit.Assert.assertEquals;

import com.loopme.request.validation.Invalidation;
import com.loopme.request.validation.Validation;
import com.loopme.request.validation.ValidationRule;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;

public class RequestValidatorTest {

    private RequestValidator validator;
    private ArrayList<Validation> dummyRules;

    @Before
    public void setUp() {
        validator = new RequestValidator();
        dummyRules = new ArrayList<>();
    }

    @Test
    public void emptyRequired() {
        String path = "app.id";
        dummyRules.add(new Validation(path, "", new ValidationRule[]{ValidationRule.REQUIRED}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(1, result.size());
        Invalidation invalidation = result.get(0);
        assertEquals(path, invalidation.getPath());
        assertEquals(ValidationRule.REQUIRED, invalidation.getRule());
    }

    @Test
    public void blankRequired() {
        String path = "app.id";
        dummyRules.add(new Validation(path, "  ", new ValidationRule[]{ValidationRule.REQUIRED}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(1, result.size());
        Invalidation invalidation = result.get(0);
        assertEquals(path, invalidation.getPath());
        assertEquals(ValidationRule.REQUIRED, invalidation.getRule());
    }

    @Test
    public void validRequired() {
        String path = "app.id";
        dummyRules.add(new Validation(path, "adasfsdfas89080", new ValidationRule[]{ValidationRule.REQUIRED}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(0, result.size());
    }

    @Test
    public void blankValue() {
        String path = "imp[].banner.w";
        dummyRules.add(new Validation(path, "  ", new ValidationRule[]{ValidationRule.REQUIRED, ValidationRule.GREATER_THEN_ZERO}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(2, result.size());
        Invalidation invalidation1 = result.get(0);
        assertEquals(path, invalidation1.getPath());
        assertEquals(ValidationRule.REQUIRED, invalidation1.getRule());
        Invalidation invalidation2 = result.get(1);
        assertEquals(path, invalidation2.getPath());
        assertEquals(ValidationRule.GREATER_THEN_ZERO, invalidation2.getRule());
    }

    @Test
    public void notGreaterThanZero() {
        String path = "imp[].banner.w";
        dummyRules.add(new Validation(path, "0", new ValidationRule[]{ValidationRule.REQUIRED, ValidationRule.GREATER_THEN_ZERO}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(1, result.size());
        Invalidation invalidation = result.get(0);
        assertEquals(path, invalidation.getPath());
        assertEquals(ValidationRule.GREATER_THEN_ZERO, invalidation.getRule());
    }

    @Test
    public void validRequiredAndGTZ() {
        String path = "imp[].banner.w";
        dummyRules.add(new Validation(path, "480", new ValidationRule[]{ValidationRule.REQUIRED, ValidationRule.GREATER_THEN_ZERO}));
        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(0, result.size());
    }

    @Test
    public void oneValidOneNot() {
        String appIdPath = "app.id";
        dummyRules.add(new Validation(appIdPath, "adasfsdfas89080", new ValidationRule[]{ValidationRule.REQUIRED}));
        String bannerWidthPath = "imp[].banner.w";
        dummyRules.add(new Validation(bannerWidthPath, "0", new ValidationRule[]{ValidationRule.REQUIRED, ValidationRule.GREATER_THEN_ZERO}));

        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(1, result.size());

        Invalidation invalidation = result.get(0);
        assertEquals(bannerWidthPath, invalidation.getPath());
        assertEquals(ValidationRule.GREATER_THEN_ZERO, invalidation.getRule());
    }

    @Test
    public void oneNotValidOneAndOneValid() {
        String bannerWidthPath = "imp[].banner.w";
        dummyRules.add(new Validation(bannerWidthPath, "0", new ValidationRule[]{ValidationRule.REQUIRED, ValidationRule.GREATER_THEN_ZERO}));
        String appIdPath = "app.id";
        dummyRules.add(new Validation(appIdPath, "adasfsdfas89080", new ValidationRule[]{ValidationRule.REQUIRED}));

        ArrayList<Invalidation> result = validator.validate(dummyRules);
        assertEquals(1, result.size());

        Invalidation invalidation = result.get(0);
        assertEquals(bannerWidthPath, invalidation.getPath());
        assertEquals(ValidationRule.GREATER_THEN_ZERO, invalidation.getRule());
    }
}

File: loopme-sdk/src/test/java/com/loopme/request/RequestValidatorTest.kt
package com.loopme.request


import com.loopme.loaders.AdRequestType
import com.loopme.request.RequestBuilder.*
import com.loopme.request.RequestValidator.Companion.BANNER_HEIGHT
import com.loopme.request.RequestValidator.Companion.BANNER_WIDTH
import com.loopme.request.RequestValidator.Companion.EVENTS_EXT_PN
import com.loopme.request.RequestValidator.Companion.EVENTS_EXT_PV
import com.loopme.request.RequestValidator.Companion.SOURCE_EXT_PN
import com.loopme.request.RequestValidator.Companion.SOURCE_EXT_PV
import com.loopme.request.RequestValidator.Companion.VIDEO_HEIGHT
import com.loopme.request.RequestValidator.Companion.VIDEO_WIDTH
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Test
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.stream.Collectors


class RequestValidatorTest {

    private val validator = RequestValidator()
    private val classLoader = ClassLoader.getSystemClassLoader()
    private val dummyAdRequestType = AdRequestType(true, true, true)

    @Test
    fun `No app object`() {
        val request = JSONObject(convertFileToString("no_app"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertTrue(violations.size == 1)
        assertEquals(RequestValidator.NOT_PRESENT, violations[APPKEY])
    }

    @Test
    fun `No appKey`() {
        val req = JSONObject(convertFileToString("no_appkey"))
        val isValid = validator.validateOrtbRequest(req, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertTrue(violations.size == 1)
        assertEquals(RequestValidator.NOT_PRESENT, violations[APPKEY])
    }

    @Test
    fun `Empty appKey`() {
        val request = JSONObject(convertFileToString("empty_appkey"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertTrue(violations.size == 1)
        assertEquals(RequestValidator.BLANK, violations[APPKEY])
    }

    @Test
    fun `No source_ext_omidpn,no source_ext_omidpv`() {
        val request = JSONObject(convertFileToString("all_required_not_present"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.NOT_PRESENT, violations[SOURCE_EXT_PN])
        assertEquals(RequestValidator.NOT_PRESENT, violations[SOURCE_EXT_PV])
    }

    @Test
    fun `Empty source_ext_omidpn and source_ext_omidpv`() {
        val request = JSONObject(convertFileToString("all_required_empty"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.BLANK, violations[SOURCE_EXT_PN])
        assertEquals(RequestValidator.BLANK, violations[SOURCE_EXT_PV])
    }

    @Test
    fun `No events_ext_omidpn,no events_ext_omidpv`() {
        val request = JSONObject(convertFileToString("all_required_not_present"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.NOT_PRESENT, violations[EVENTS_EXT_PN])
        assertEquals(RequestValidator.NOT_PRESENT, violations[EVENTS_EXT_PV])
    }

    @Test
    fun `Empty events_ext_omidpn and events_ext_omidpv`() {
        val request = JSONObject(convertFileToString("all_required_empty"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.BLANK, violations[EVENTS_EXT_PN])
        assertEquals(RequestValidator.BLANK, violations[EVENTS_EXT_PV])
    }

    @Test
    fun `Banner width and height zero`() {
        val request = JSONObject(convertFileToString("all_required_empty"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.VALUE_0, violations[BANNER_WIDTH])
        assertEquals(RequestValidator.VALUE_0, violations[BANNER_HEIGHT])
    }

    @Test
    fun `Banner width and height not present`() {
        val request = JSONObject(convertFileToString("all_required_not_present"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[BANNER_WIDTH])
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[BANNER_HEIGHT])
    }

    @Test
    fun `Video width and height invalid value`() {
        val request = JSONObject(convertFileToString("all_required_empty"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType.copy())
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[VIDEO_WIDTH])
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[VIDEO_HEIGHT])
    }

    @Test
    fun `Video width and height not present`() {
        val request = JSONObject(convertFileToString("all_required_not_present"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType.copy())
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[VIDEO_WIDTH])
        assertEquals(RequestValidator.NOT_PRESENT_OR_INVALID, violations[VIDEO_HEIGHT])
    }

    @Test
    fun `Banner only when is not a video not rewarded`() {
        val request = JSONObject(convertFileToString("banner_only"))
        val adRequestType = dummyAdRequestType.copy(isBanner = true, isVideo = false, isRewarded = false)
        val isValid = validator.validateOrtbRequest(request, adRequestType)
        val violations = validator.violations
        assertFalse(isValid)
        assertEquals(RequestValidator.VALUE_0, violations[BANNER_WIDTH])
        assertEquals(RequestValidator.VALUE_0, violations[BANNER_HEIGHT])
        assertFalse(violations.containsKey(VIDEO_HEIGHT))
        assertFalse(violations.containsKey(VIDEO_WIDTH))
    }

    @Test
    fun `Is valid for banner and video`() {
        val request = JSONObject(convertFileToString("proper_ad_request"))
        val isValid = validator.validateOrtbRequest(request, dummyAdRequestType)
        val violations = validator.violations
        assertTrue(isValid)
        assertEquals(0, violations.size)
    }

    private fun convertFileToString(fileName: String) =
        BufferedReader(
            InputStreamReader(
                classLoader.getResourceAsStream(fileName)
            )
        ).lines().collect(Collectors.joining("\n"))
}